databuilder/databuilder/publisher/neo4j_csv_publisher_apoc.py

```diff
@@ -0,0 +1,456 @@
+                LOGGER.info('_publish_node(%s)', node_file)
+                quick = time.time()
+                self._publish_node(node_file)
+                LOGGER.info('Successfully published nodes. Elapsed: %i seconds', time.time() - quick)
+            except StopIteration:
+                break
+
+        LOGGER.info('Publishing Relationship files: %s', self._relation_files)
+        while True:
+            try:
+                relation_file = next(self._relation_files_iter)
+                LOGGER.info('_publish_relation(%s)', relation_file)
+                quick = time.time()
+                self._publish_relation(relation_file)
+                LOGGER.info('Successfully published relationships. Elapsed: %i seconds', time.time() - quick)
+            except StopIteration:
+                break
+
+        # TODO: Add statsd support
+        LOGGER.info('Successfully published all. Elapsed: %i seconds', time.time() - start)
+
+    def get_scope(self) -> str:
+        return 'publisher.neo4j'
+
+    def _create_indices(self, node_file: str) -> None:
+        """
+        Go over the node file and try creating unique index
+        :param node_file:
+        :return:
+        """
+        LOGGER.info('Creating indices. (Existing indices will be ignored)')
+
+        with open(node_file, 'r', encoding='utf8') as node_csv:
+            for node_record in pandas.read_csv(node_csv, na_filter=False).to_dict(orient='records'):
+                label = node_record[NODE_LABEL_KEY]
+                if label not in self.labels:
+                    self._try_create_index(label)
+                    self.labels.add(label)
+
+        LOGGER.info('Indices have been created.')
+
+    def _chunker(self, seq, size):
+        return (seq[pos:pos + size] for pos in range(0, len(seq), size))
+
+    def _unquote(self, k) -> str:
+        if k.endswith(UNQUOTED_SUFFIX):
+            k = k[:-len(UNQUOTED_SUFFIX)]
+        return k
+
+    def _publish_node(self, node_file: str) -> None:
+        """
+        Iterate over the csv records of a file, each csv record transform to apoc.node.merge statement and will be executed.
+        All nodes should have a unique key, and this method will try to create unique index on the LABEL when it sees
+        first time within a job scope.
+        Example of Cypher query executed by this method:
+        CALL apoc.periodic.iterate(
+            'UNWIND $rows AS row RETURN row', 
+            'CALL apoc.merge.node([row.label], {key:row.key}, row, row) YIELD node RETURN COUNT(*);', 
+            {batchSize: $batch_size, iterateList: True, parallel: False, params: { rows: $batch, tag: $publish_tag }}
+        );
+        :param node_file:
+        :return:
+        """
+        df = pandas.read_csv(node_file, na_filter=False)
+        df = df.rename(columns=self._unquote) # Remove the UNQUOTED_SUFFIX+
+        df.columns= df.columns.str.lower() # Lowercase the columns
```

Comment:
```suggestion
        
```
Suggestion to remove this line of code as it causes `KeyError` for `LABEL` (it becomes `label` after this line here) when specifying nodes that are create only (`CREATE_ONLY_NODES`)

See line 294 for the point of failure.

Reply:
Not sure if the `UNQUOTED_SUFFIX+` will raise issues as well. For `node_record`, it expects:
```python
{
  'LABEL': 'Table', 
  'KEY': 'vega://vega.curated_analytics_green/cur_m_m_merchant_master_with_gross_rev', 
  'name': 'cur_m_m_merchant_master_with_gross_rev', 
  'is_view:UNQUOTED': False
}
```

But this version will remove it resulting in:
```python
{
  'label': 'Table', 
  'key': 'vega://vega.curated_analytics_green/cur_m_m_merchant_master_with_gross_rev', 
  'name': 'cur_m_m_merchant_master_with_gross_rev', 
  'is_view': False, 
  'published_tag': '2022-04-29', 
  'publisher_last_updated_epoch_ms': 1651192078253176
}
```

It might result in an issue if someone specifies `is_view` (haven't tested it)

Reply:
@akiratwang I left that logic in because it looked like that was happening in _create_props_param() in the original version.

Reply:
@akiratwang Removed the lower(). There is also a lower() on line 363 in the _publish_relation(). Should we remove that too? I honestly can't remember why I added those in the first place.

Reply:
Hmmm I guess if we remove it and it works why not hahaha

I didn't pick up the one on line 363 but that might help with some nodes being missing (i.e adding several programmatic descriptions) during the bulk loads.

Reply:
This is what I mean with multiple description sources: https://github.com/amundsen-io/amundsen/discussions/1828

But it's such a niche "want" or "feature" so I'm doubtful hahaha

Reply:
Done!